File: 윤장원/210630/18년 카카오공채/셔틀버스.py
def solution(n, t, m, timetable):
    timetable = [ int(time[:2]) * 60 + int(time[3:]) for time in timetable]
    timetable.sort()
    
    current = 540
    
    for _ in range(n):
        for _ in range(m):
            # 대기가 있는 경우 1분 전 도착
            if timetable and timetable[0] <= current:
                candidate = timetable.pop(0) - 1
            else:
                candidate = current
                
        current += t
        
    # 시, 분으로 변경
    h, m = divmod(candidate, 60)
    return str(h).zfill(2) + ':' + str(m).zfill(2)
    
    
    # 셔틀은 09:00부터 총 n회 t분 간격으로 역에 도착하며, 하나의 셔틀에는 최대 m명의 승객이 탈 수 있다.

File: 윤장원/210630/18년 카카오공채/셔틀버스.js
// 풀이 1
function solution(n, t, m, timetable) {
    const timetables = [];
    for (let time of timetable) {
        time = time.split(':').map((x) => +x);
        const [hour, min] = time;
        timetables.push(hour * 60 + min);
    }
    
    timetables.sort((a,b) => a - b);
    let current = 540;
    let candidate = 0;
    
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            if (timetables && timetables[0] <= current) {
                candidate = timetables.shift() - 1;
            } else {
                candidate = current;
            }
        }
        current += t
    }
    
    let hour = String(parseInt(candidate / 60));
    let min = String(candidate % 60);
    
    if (hour.length === 1) {
        hour = '0' + hour;
    }
    
    if (min.length === 1) {
        min = '0' + min;
    }
    
    return hour + ':' + min;
}


// 풀이 2
function solution(n, t, m, timetable) {
    const getTime = time => (+time.substring(0,2) * 60) + (+time.substring(3));
    
    timetable = timetable.map((x) => x = getTime(x)).sort((a, b) => a - b);
    let current = 540;
    let candidate = 0;
    
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            if (timetable && timetable[0] <= current) {
                candidate = timetable.shift() - 1;
            } else {
                candidate = current;
            }
        }
        current += t
    }
    
    let hour = String(parseInt(candidate / 60));
    let min = String(candidate % 60);
    
    if (hour.length === 1) {
        hour = '0' + hour;
    }
    
    if (min.length === 1) {
        min = '0' + min;
    }
    
    return hour + ':' + min;
}

/*

    1) n -> 횟수
    2) t -> 분
    3) m -> 최대 사람 승객
    4) 대기시간 - 1
*/